src/mlpack/core/cv/k_fold_cv.hpp

```diff
@@ -0,0 +1,219 @@
+   *     constructors for reference).
+   */
+  template<typename... CVBaseArgs>
+  KFoldCV(const size_t k, const CVBaseArgs&... args);
+
+  /**
+   * Run k-fold cross-validation.
+   *
+   * @param args Arguments for MLAlgorithm (in addition to the passed
+   *     ones in the constructor).
+   */
+  template<typename... MLAlgorithmArgs>
+  double Evaluate(const MLAlgorithmArgs& ...args);
+
+  //! Access and modify a model from the last run of k-fold cross-validation.
+  MLAlgorithm& Model();
+
+ private:
+  //! A short alias for CVBase.
+  using Base = CVBase<MLAlgorithm, MatType, PredictionsType, WeightsType>;
+
+  //! The number of bins in the dataset.
+  const size_t k;
+
+  /**
+   * Variables for storing the extended (by repeating the first k - 2 bins)
+   * dataset.
+   */
+  MatType xs;
+  PredictionsType ys;
+  WeightsType weights;
+
+  //! The size of each bin in terms of data points.
+  size_t binSize;
+
+  //! The size of each training subset in terms of data points.
+  size_t trainingSubsetSize;
+
+  //! A pointer to a model from the last run of k-fold cross-validation.
+  std::unique_ptr<MLAlgorithm> modelPtr;
+
+  /**
+   * Initialize without weights.
+   */
+  template<typename DataArgsTupleT,
+           typename = typename std::enable_if<
+               std::tuple_size<DataArgsTupleT>::value == 2>::type>
+  void Init(const DataArgsTupleT& dataArgsTuple);
+
+  /**
+   * Initialize with weights.
+   */
+  template<typename DataArgsTupleT,
+           typename = typename std::enable_if<
+               std::tuple_size<DataArgsTupleT>::value == 3>::type,
+           typename = void>
+  void Init(const DataArgsTupleT& dataArgsTuple);
+
+  /**
+   * Initialize the given destination matrix with the given source joined with
+   * its first k - 2 bins.
+   */
+  template<typename SourceType, typename DestinationType>
+  void InitKFoldCVMat(const SourceType& source, DestinationType& destination);
+
+  /**
+   * Train and run evaluation in the case of non-weighted learning.
+   */
+  template<typename...MLAlgorithmArgs,
+           bool Enabled = !Base::MIE::SupportsWeights,
+           typename = typename std::enable_if<Enabled>::type>
+  double TrainAndEvaluate(const MLAlgorithmArgs& ...mlAlgorithmArgs);
+
+  /**
+   * Train and run evaluation in the case of supporting weighted learning.
+   */
+  template<typename...MLAlgorithmArgs,
+           bool Enabled = Base::MIE::SupportsWeights,
+           typename = typename std::enable_if<Enabled>::type,
+           typename = void>
+  double TrainAndEvaluate(const MLAlgorithmArgs& ...mlAlgorithmArgs);
+
+  /**
+   * Calculate the index of the first column of the ith validation subset.
+   *
+   * We take the ith validation subset after the ith training subset if
+   * i < k - 1 and before it otherwise.
+   */
+  size_t ValidationSubsetFirtsCol(const size_t i)
+  {
+    return i < k - 1? binSize * i + trainingSubsetSize : binSize * (i - 1);
```

Comment:
Minor style comment---can you add a space between the '1' and '?'?

Personally I like to make the ternary operands more "grouped" by grouping them
in parentheses, like this:

  return (i < k - 1) ? (binSize * i + trainingSubsetSize) : (binSize * (i - 1));

I think it's a little easier to read like that.  However it is up to you whether
or not you want to group them like that. :)